src/components/tl/mlx5/mcast/tl_mlx5_mcast_rcache.c

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_mlx5_mcast_rcache.h"
+
+static ucs_status_t ucc_tl_mlx5_mcast_coll_reg_mr(ucc_tl_mlx5_mcast_coll_context_t
+                                                 *ctx, void *data, size_t data_size,
+                                                  void **mr)
+{
+    *mr = ibv_reg_mr(ctx->pd, data, data_size, IBV_ACCESS_LOCAL_WRITE |
+                     IBV_ACCESS_REMOTE_READ | IBV_ACCESS_REMOTE_WRITE);
+
+    tl_trace(ctx->lib, "external memory register: ptr %p, len %zd, mr %p",
+             data,  data_size, (*mr));
+    if (!*mr) {
+        tl_error(ctx->lib, "failed to register MR\n");
+        return UCS_ERR_NO_MEMORY;
+    }
+    
+    return UCS_OK;
+}
+
+
+static ucc_status_t ucc_tl_mlx5_mcast_coll_dereg_mr(ucc_tl_mlx5_mcast_coll_context_t
+                                                   *ctx, void *mr)
+{
+    if(mr != NULL) {
+        tl_debug(ctx->lib, "external memory deregister: mr %p", mr);
+        if (ibv_dereg_mr(mr)) {
+            tl_error(ctx->lib, "couldn't destroy mr %p", mr);
+            return UCC_ERR_NO_RESOURCE;
+        }
+    } else {
+        tl_debug(ctx->lib, "external memory mr %p already deregistered", mr);
+    }
+
+    return UCC_OK;
+}
+
+static ucs_status_t
+ucc_tl_mlx5_mcast_rcache_mem_reg_cb(void *context, ucs_rcache_t *rcache,
+                                    void *arg, ucs_rcache_region_t *rregion,
+                                    uint16_t flags)
+{
+    ucc_tl_mlx5_mcast_rcache_region_t *region;
+    void                              *address;
+    size_t                             length;
+    ucs_status_t                       ret;
+
+    address = (void*)rregion->super.start;
+    length  = (size_t)(rregion->super.end - rregion->super.start);
+    region  = ucc_derived_of(rregion, ucc_tl_mlx5_mcast_rcache_region_t);
+
+    ret = ucc_tl_mlx5_mcast_coll_reg_mr((ucc_tl_mlx5_mcast_coll_context_t *)context,
+                                         address, length, &region->reg.mr);
+    if (UCS_OK != ret) {
+        return ret;
+    }
+
+    return UCS_OK;
+
+}
+
+static void ucc_tl_mlx5_mcast_rcache_mem_dereg_cb(void *context, ucc_rcache_t
+                                                 *rcache, ucc_rcache_region_t *rregion)
+{
+    ucc_tl_mlx5_mcast_rcache_region_t *region = ucc_derived_of(rregion,
+                                                ucc_tl_mlx5_mcast_rcache_region_t);
+
+    ucc_tl_mlx5_mcast_coll_dereg_mr((ucc_tl_mlx5_mcast_coll_context_t *)context,
```

Comment:
shouldn't we check error status here ?

Reply:
The only use case for checking the output of ucc_tl_mlx5_mcast_coll_dereg_mr () is to print an error. However,  if there is an error, ucc_tl_mlx5_mcast_coll_dereg_mr() is already printing the error message. 
Since ucc_tl_mlx5_mcast_rcache_mem_dereg_cb() is not returning anything, for now, we do need to check the output of the ucc_tl_mlx5_mcast_coll_dereg_mr(). If in the future, this callback is changed, we might have to check the output of dereg function as well, but for now, it is not necessary.

Reply:
got it, thanks